### ISSUE TITLE
`snprintf` tests assume 64-bit long integer

### ISSUE THREAD

**Issue by khardix** [newsboat/newsboat#555]
**Newsboat version**: 2.16.1

Several tests for both C++ and Rust wrappers of `snprintf` functions assume that `long int` can hold 64-bit values. This is not true on 32-bit architectures (as far as I can tell), where it usually seems that both `int` and `long int` are 32-bit numbers. This causes the test suite for C++ to fail to build (with overflow warning turned to error), and the Rust test suite to fail.

#### Steps to reproduce the issue:

1.  Build the tests on 32-bit arch: `$ make test`

    ```
    test/strprintf.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____17()':
    test/strprintf.cpp:137:40: error: integer overflow in expression of type 'long int' results in '2147483647' [-Werror=overflow]
      137 |  REQUIRE(strprintf::fmt("%li", int_min - 1l) == "-2147483649");
          |                                ~~~~~~~~^~~~
    test/strprintf.cpp:143:40: error: integer overflow in expression of type 'long int' results in '-2147483648' [-Werror=overflow]
      143 |  REQUIRE(strprintf::fmt("%li", int_max + 1l) == "2147483648");
          |                                ~~~~~~~~^~~~
    cc1plus: all warnings being treated as errors
    ```

2.  Run the Rust test suite on 32-bit architecture: `$ cargo test`

    ```
    ---- tests::formats_i64 stdout ----
    thread 'tests::formats_i64' panicked at 'assertion failed: `(left == right)`
      left: `"2147483647"`,
    right: `"-2147483649"`', rust/strprintf/src/lib.rs:159:9
    note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

    ---- tests::formats_u64 stdout ----
    thread 'tests::formats_u64' panicked at 'assertion failed: `(left == right)`
      left: `"4294967295"`,
    right: `"18446744073709551615"`', rust/strprintf/src/lib.rs:171:9
    ```

#### Other info you think is relevant:

The test failures were discovered when I started building v2.16.1 for Fedora. Before this version, the test suite was neither built nor run. [Latest failed build](https://koji.fedoraproject.org/koji/taskinfo?taskID=35865551).

I was able to work around/disable the relevant tests with the following patch (not a fix, but may be useful as a start for the real one):

```patch
From 0a12f7c1102789f08d3afca882906f8b0684ebb1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jan=20Stan=C4=9Bk?= <jstanek@redhat.com>
Date: Fri, 28 Jun 2019 10:50:07 +0200
Subject: [PATCH] Tweak tests for 32-bit architectures
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Jan Staněk <jstanek@redhat.com>
---
 rust/strprintf/src/lib.rs | 10 +++++++---
 test/strprintf.cpp        | 10 +++++++---
 2 files changed, 14 insertions(+), 6 deletions(-)

diff --git a/rust/strprintf/src/lib.rs b/rust/strprintf/src/lib.rs
index eb8c5d57..aa839be7 100644
--- a/rust/strprintf/src/lib.rs
+++ b/rust/strprintf/src/lib.rs
@@ -156,8 +156,10 @@ mod tests {
     #[test]
     fn formats_i64() {
         assert_eq!(fmt!("%li", 42i64), "42");
-        assert_eq!(fmt!("%li", std::i32::MIN as i64 - 1), "-2147483649");
-        assert_eq!(fmt!("%li", std::i32::MAX as i64 + 1), "2147483648");
+        if cfg!(target_pointer_width = "64") {
+            assert_eq!(fmt!("%li", std::i32::MIN as i64 - 1), "-2147483649");
+            assert_eq!(fmt!("%li", std::i32::MAX as i64 + 1), "2147483648");
+        };

         assert_eq!(fmt!("%lli", 42i64), "42");
         assert_eq!(fmt!("%lli", std::i64::MIN), "-9223372036854775808");
@@ -168,7 +170,9 @@ mod tests {
     fn formats_u64() {
         assert_eq!(fmt!("%lu", 42u64), "42");
         assert_eq!(fmt!("%lu", 0u64), "0");
-        assert_eq!(fmt!("%lu", std::u64::MAX), "18446744073709551615");
+        if cfg!(target_pointer_width = "64") {
+            assert_eq!(fmt!("%lu", std::u64::MAX), "18446744073709551615");
+        };

         assert_eq!(fmt!("%llu", 42u64), "42");
         assert_eq!(fmt!("%llu", 0u64), "0");
diff --git a/test/strprintf.cpp b/test/strprintf.cpp
index 2e57bdd2..591339ef 100644
--- a/test/strprintf.cpp
+++ b/test/strprintf.cpp
@@ -1,5 +1,6 @@
 #include "strprintf.h"

+#include <climits>
 #include <limits>

 #include "3rd-party/catch.hpp"
@@ -130,17 +131,19 @@ TEST_CASE("strprintf::fmt() formats long int", "[strprintf]")
 {
 	REQUIRE(strprintf::fmt("%li", 42l) == "42");

+#if LONG_MIN < INT_MIN
 	const auto int_min = std::numeric_limits<int>::min();
 	const auto long_min = std::numeric_limits<long int>::min();
-	REQUIRE(long_min < int_min);
 	REQUIRE(int_min == -2147483648);
 	REQUIRE(strprintf::fmt("%li", int_min - 1l) == "-2147483649");
+#endif

+#if LONG_MAX > INT_MAX
 	const auto int_max = std::numeric_limits<int>::max();
 	const auto long_max = std::numeric_limits<long int>::max();
-	REQUIRE(long_max > int_max);
 	REQUIRE(int_max == 2147483647);
 	REQUIRE(strprintf::fmt("%li", int_max + 1l) == "2147483648");
+#endif
 }

 TEST_CASE("strprintf::fmt() formats long unsigned int", "[strprintf]")
@@ -149,9 +152,10 @@ TEST_CASE("strprintf::fmt() formats long unsigned int", "[strprintf]")

 	REQUIRE(strprintf::fmt("%lu", 0lu) == "0");

+#if ULONG_MAX == 18446744073709551615lu
 	const auto ulong_max = std::numeric_limits<long unsigned int>::max();
-	REQUIRE(ulong_max == 18446744073709551615lu);
 	REQUIRE(strprintf::fmt("%lu", ulong_max) == "18446744073709551615");
+#endif
 }

 TEST_CASE("strprintf::fmt() formats long long int", "[strprintf]")
--
2.21.0
```

**Comment by khardix** [comment]
Bonus round: On ARM (`armv7hl`), even the `42i64` and `42u64` formatting rust tests fail:

```
---- tests::formats_u64 stdout ----
thread 'tests::formats_u64' panicked at 'assertion failed: `(left == right)`
  left: `"3057647632"`,
 right: `"42"`', rust/strprintf/src/lib.rs:169:9
---- tests::formats_i64 stdout ----
thread 'tests::formats_i64' panicked at 'assertion failed: `(left == right)`
  left: `"-1244659696"`,
 right: `"42"`', rust/strprintf/src/lib.rs:158:9
```

I ended up disabling these test entirely on 32-bit platforms (`#[cfg(target_pointer_width = "64")]`).

**Comment by Minoru** [comment]
Hi! Thank you for the report. I'll add i386 to our continuous integration pipeline (#556), and then work on a proper fix for this.

**Comment by Minoru** [comment]
We now have an i686 job on Cirrus CI, so can work on fixing this. Here's a brain dump of what I plan to do about this bug.

1. Rust tests fail because on i686, `snprintf` doesn't accept 64-bit values (or something like that). The question then is: do we even need them?

    **Actionable:** in C++, remove `fmt` overloads for `long` and `long long`, and see what breaks (both on i686 and x86_64).

2. I believe our C++ sometimes uses `int`s to hold indexes into collections, and commits other minor sins like this. The compiler might let this slide, because implicit conversions. Rust is more stringent about this, and will force us to choose between `i32`, `i64`, and `isize`. Maybe try to force the issue, and see if we can make `fmt` use fixed-width types only?

    **Actionable:** in C++, replace all integer overloads of `fmt` with `fmt(int32_t)` and `fmt(uint32_t)`, see how much stuff breaks and, if possible, list common mistakes. (My expectation is that the only problems will be `size_t`, `ssize_t`, and `time_t`. All three can be casted to 32-bit type, because it's highly unlikely that anyone has that many items/feeds/loop iterations, and time_t will fit into 32 bit until 2038. Plus this only impacts formatting, which is mostly logging.)

And we probably need a CI job for ARM as well.

**Comment by khardix** [comment]
OK, so here is my brain dump 🙂. The TL;DR version is that there should be no need to sacrifice 64-bit values on 32-bit architectures, it only requires more careful/precise handling of C integral types.

Long version/reply:

1.  > Rust tests fail because on i686, snprintf doesn't accept 64-bit values (or something like that).

    Not quite correct. The trouble is that in C (and so in C++), the "native" (`long`, `short`, `int`) integer types have implementation-defined width; the standard only requires that `sizeof(short) <= sizeof(int)` and `sizeof(int) <= sizeof(long)`. In theory it is possible to have a compiler/architecture where all these integers are of the same width (although I'm not aware of any compiler actually doing this). The issue here is the *assumption* that `sizeof(long) == 8  // 64bits` on all arches.

    The good news is that even on 32-bit architecture, there should be a `int64_t` type (see below for quick PoC).

    As a first step in Rust compatibility here, I would suggest to gradually move to fixed-size integers, and using [macros for formatting strings](https://en.cppreference.com/w/cpp/types/integer#Format_macro_constants):

    ```c++
    #include <cinttypes>
    #include <cstdint>
    #include <cstdio>

    std::uint64_t answer = 42;  // bare uint64_t may also work
    std::printf("%" PRIu64 "\n", answer);
    ```

    Alternatively, use diferrent abstraction for formatting than printf-style, since that is not really compatible with the rust way of doing things.

2.  > In C++, replace all integer overloads of fmt with fmt(int32_t) and fmt(uint32_t), see how much stuff breaks and, if possible, list common mistakes.

    Well, here I would provide overrides for `int32_t`, `uint32_t`, `int64_t` and `uint64_t` (or even just the 64-bit variants), and then let the `size_t` and friends use whatever type they are defined as.

---

The short C11 program I have used for testing type widths (compiled on Fedora x86_64, so the 32-bit one is cross-compiled and probably should also be tested in pure 32-bit environment):

```c
#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>

int main() {
    uint64_t answer = 42;
    printf("%s: ", PRIu64);
    printf("%" PRIu64 "\n", answer);
    return 0;
}
```

-   64-bit (`$ gcc -std=c11 width-test.c && file a.out && ./a.out`):

        a.out: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), ...
        lu: 42

-   32-bit (`$ gcc -m32 -std=c11 width-test.c && file a.out && ./a.out`):

        a.out: ELF 32-bit LSB executable, Intel 80386, version 1 (SYSV), ...
        llu: 42

**Comment by Minoru** [comment]
Yeah, it looks like the solution would be to gradually move to fixed-width types.

> Alternatively, use diferrent abstraction for formatting than printf-style, since that is not really compatible with the rust way of doing things.

That's the goal, but as long as we have both C++ and Rust code, we need identical formatting facilities in both. We already had an `snprintf` wrapper in C++, so porting that to Rust seemed like the best way forward.

> Well, here I would provide overrides for int32_t, uint32_t, int64_t and uint64_t (or even just the 64-bit variants), and then let the size_t and friends use whatever type they are defined as.

I like this idea. **Actionable**: check if libc crate provides `PRI???` macros, and figure out how to use them from Rust.

**Comment by Minoru** [comment]
libc crate doesn't have `PRI*` constants. avr-libc does, but it also pulls other dependencies that we don't need. Since we only need like six constants, I'd just create two new crates in Newsboat, with definitions for 32-bit and 64-bit architectures, and be done with this. They should be defined as `&str`, not `[u8; 2]` as avr-libc does it, because the constants will be concatenated to `String` and `&str` anyway.

> Well, here I would provide overrides for int32_t, uint32_t, int64_t and uint64_t (or even just the 64-bit variants), and then let the size_t and friends use whatever type they are defined as.

This didn't work out: in our i686 container, `time_t` is defined as `long int`, while `int32_t` is `int` and `int64_t` is `long long int`.

On the other hand, not that much stuff broke when I removed old `fmt` overloads, so I might be able to just fix everything to use fixed-width types.

**Comment by Minoru** [comment]
I fixed the x86 part in #590.

ARM errors puzzle me; I don't understand how they can even happen. Frankly, I'm a bit tired of wrangling with portability problems, so I'll lay that one aside for some time.
